src/Workspaces/Core/Portable/Rename/Renamer.RenameSymbolDocumentAction.cs

```diff
@@ -0,0 +1,152 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Immutable;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Resources;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Rename.ConflictEngine;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.Rename
+{
+    public static partial class Renamer
+    {
+        /// <summary>
+        /// Action that will rename a type to match the current document name
+        /// </summary>
+        internal class RenameSymbolDocumentAction : RenameDocumentAction
+        {
+            private readonly RenameSymbolDocumentActionAnalysis _analysis;
+
+            private RenameSymbolDocumentAction(
+                Document document,
+                RenameSymbolDocumentActionAnalysis analysis,
+                OptionSet optionSet,
+                ImmutableArray<string> errors)
+                : base(document.Id, optionSet, errors)
+            {
+                _analysis = analysis;
+            }
+
+            public override string GetDescription(CultureInfo? culture)
+            => string.Format(WorkspacesResources.ResourceManager.GetString("Rename_0_to_1", culture ?? WorkspacesResources.Culture), _analysis.OriginalDocumentName, _analysis.NewDocumentName);
+
+            public override async Task<Solution> GetModifiedSolutionAsync(Solution solution, CancellationToken cancellationToken = default)
+            {
+                // Always make sure the document name is correctly updated
+                solution = solution.WithDocumentName(DocumentId, _analysis.NewDocumentName);
```

Comment:
Open question: 

For calls to `Solution.WithDocumentXXX` should they be a separate action?

Reply:
I guess it seems like this should be happening in the RenameDocumentActionInfo GetSolutionAsync or something: you don't want _each_ action to have to do this if there's multiple actions involved, right?